_drafts/zero-to-main-1.md

```diff
@@ -0,0 +1,328 @@
+            <td>[Initialize any runtime]</td>
+            <td>Optionally make calls to C/C++ runtime init code to enable use
+of heap, floating point, or other features. This is normally done by <code
+class="literal">__main</code> from the C/C++ library.</td>
+        </tr>
+    </tbody>
+</table>
+<br/>
+So, our ResetHandler is responsible for initializing static and global
+variables, and starting our program.
+This mirrors what the C standards tells us:
+
+> All objects with static storage duration shall be initialized (set to their
+> initial values) before program startup. The manner and timing of such
+> initialization are otherwise unspecified. 
+
+(Section 5.1.2, Execution environment)
+
+In practice this means that given the following snippet:
+
+```c
+static uint32_t foo;
+static uint32_t bar = 2;
+```
+
+Our `Reset_Handler` needs to make sure that the memory at `&foo` is 0x00000000, and the
+memory at `&bar` is 0x00000002.
+
+We cannot just go and initialize each variable one by one. Instead, we rely on
+the compiler (technically, the linker) to put all those variables in the same
+place so we can initialize them in one fell swoop.
+
+For static variables that must be zeroed, the linker gives us `_sbss` and
+`_ebss` as start and end addresses. We can therefore do:
+
+```c
+/* Clear the zero segment */
+for (uint32_t *bss_ptr = &_sbss; bss_ptr < &_ebss;) {
+    *bss_ptr++ = 0;
+}
+```
+
+For static variables with an init value, the linker gives us:
+* `_etext` as the address the init values are stored at
+* `_sdata` as the address the static variables live at
+* `_edata` as the end of the static variables memory
+
+We then can do:
+```c
+uint32_t *init_values_ptr = &_etext;
+uint32_t *data_ptr = &_sdata;
+if (init_values_ptr != data_ptr) {
+    for (; data_ptr < &_edata;) {
+        *data_ptr++ = *init_values_ptr++;
+    }
+}
+```
+
+Putting it together, we can write our `Reset_Handler`
+
+```c
+void Reset_Handler(void)
+{
+    /* Copy init values from text to data */
+    uint32_t *init_values_ptr = &_etext;
+    uint32_t *data_ptr = &_sdata;
+
+    if (init_values_ptr != data_ptr) {
+        for (; data_ptr < &_edata;) {
+            *data_ptr++ = *init_values_ptr++;
+        }
+    }
+
+    /* Clear the zero segment */
+    for (uint32_t *bss_ptr = &_sbss; bss_ptr < &_ebss;) {
+        *bss_ptr++ = 0;
+    }
+}
+```
+
+We still need to start our program! That's achieved with a simple call to
+`main()`.
+
+```c
+void Reset_Handler(void)
+{
+    /* Copy init values from text to data */
+    uint32_t *init_values_ptr = &_etext;
+    uint32_t *data_ptr = &_sdata;
+
+    if (init_values_ptr != data_ptr) {
+        for (; data_ptr < &_edata;) {
+            *data_ptr++ = *init_values_ptr++;
+        }
+    }
+
+    /* Clear the zero segment */
+    for (uint32_t *bss_ptr = &_sbss; bss_ptr < &_ebss;) {
+        *bss_ptr++ = 0;
+    }
+
+    /* Overwriting the default value of the NVMCTRL.CTRLB.MANW bit
+       (rata reference 13134) */
```

Comment:
Is "rata" a thing? Should it be "errata"?

Reply:
typo! Fixed.